hal/ia32/pci.c

```diff
@@ -106,11 +111,64 @@ int hal_pciSetBusmaster(pci_dev_t *dev, u8 enable)
 }
 
 
+int hal_pciSetUsbOwnership(pci_dev_t *dev, u8 eecp, u8 osOwned)
+{
+	spinlock_ctx_t sc;
+	u32 dv;
+	u8 reg = eecp >> 2U; /* eecp is a pci config offset */
+
+	if (dev == NULL) {
+		return -EINVAL;
+	}
+
+	hal_spinlockSet(&pci_common.spinlock, &sc);
+	dv = _hal_pciGet(dev->bus, dev->dev, dev->func, reg);
+	if (osOwned != 0) {
+		dv |= (1 << 24);
+	}
+	else {
+		dv &= ~(1 << 24);
+	}
+	_hal_pciSet(dev->bus, dev->dev, dev->func, reg, dv);
+
+	for (;;) {
+		dv = _hal_pciGet(dev->bus, dev->dev, dev->func, reg);
+		if ((osOwned != 0) && ((dv & (1 << 24)) != 0) && ((dv & (1 << 16)) == 0)) {
+			break;
+		}
+
+		if ((osOwned == 0) && ((dv & (1 << 24)) == 0) && ((dv & (1 << 16)) != 0)) {
+			break;
+		}
```

Comment:
```suggestion
```

Reply:
These two conditionals are for handling two separate situations: first one is relevant when the OS asks the BIOS for ownership of the USB controller, second one is when the OS wants to relinquish the ownership and return it to the BIOS. These conditions check for the states of "HC BIOS Owned" and "HC OS Owned" semaphores, respectively, to see whether the ownership was obtained/moved as needed. Removing the second case would hang the caller indefinitely if `osOwned == 0`